[jit] Add interface/object serialization as module attribute

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27236 [WIP] module swapping as attribute in recursive API
* **#26770 [jit] Add interface/object serialization as module attribute**

Summary:

This PR added the run-time class serialization to allow interface/class
polymorphism, it modifes the Pickler to keep a list of run-time class
types during pickling the IValues, then serialize all these types in export.cpp.

e.g. to allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26770

Differential Revision: [D17742707](https://our.internmc.facebook.com/intern/diff/D17742707)

wanchaol committed:
[jit] Add interface/object serialization as module attribute

Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -550,6 +553,24 @@ class ScriptModuleSerializer {
   }
 
  private:
+  void writeMaterializedInterfaceInstances(const script::Module& module) {
```
This approach creates a second pass over the object hierarchy that can easily get out of sync with the pickler. It already is out of sync: what about tuples? what about nested modules? The pickler should be modified to keep a list of all the types it wrote, and that should be added to the list of types that writeCode writes.

zdevito reviewed (changes requested):
Approach here needs to be based on the pickler or it will be very fragile.

wanchaol committed:
address comments on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
rebase and address comments on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito commented on torch/csrc/jit/export.cpp:
```diff
@@ -577,6 +577,11 @@ class ScriptModuleSerializer {
     std::stringstream fname;
     fname << archive_name << ".pkl";
     writer_.writeRecord(fname.str(), data.data(), data.size());
+    // write all the captured run-time class types during pickling the IValues
```
nit: This setup is slightly different than the way the python printer works. The python printer takes a reference to a vector and appends the class dependencies. Can we make the pickler do the same? When I land https://github.com/pytorch/pytorch/pull/26788 I am going to have reconcile to two and it would be better if they both could write to the class_deps_ array.

zdevito reviewed (approved):
Looks good, one minor comment.

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
rebase on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the interface/object serialization as module attribute, to
allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17742707](https://our.internmc.facebook.com/intern/diff/D17742707)

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the run-time class serialization to allow interface/class
polymorphism, it modifes the Pickler to keep a list of run-time class
types during pickling the IValues, then serialize all these types in export.cpp.

e.g. to allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26770

Differential Revision: [D17742707](https://our.internmc.facebook.com/intern/diff/D17742707)

wanchaol committed:
Update on "[jit] Add interface/object serialization as module attribute"


Summary:

This PR added the run-time class serialization to allow interface/class
polymorphism, it modifes the Pickler to keep a list of run-time class
types during pickling the IValues, then serialize all these types in export.cpp.

e.g. to allow initializing object as a interface type during python
initialization. Because interface type can be backed by any class object
that implements that interface, if we declare it in
python/module.__init__, we will need to collect the run time types of the
value and serialize them to ensure complete code information

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/26770

Differential Revision: [D17742707](https://our.internmc.facebook.com/intern/diff/D17742707)
